### PR TITLE
feat(core): update password in profile

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Implement API `POST /profile/password` for changing password. Include the use of verification record and end user's oidc access token.

Reuse the class `PasswordVerification` with some changes:

1. Extend `identifier` to accept `userId` as an identifier

Please notice that the class `PasswordVerification` is still in the old directory for better code review, will move them to another folder in LOG-9983

## Testing

Integration tests.

## Checklist

- [ ] `.changeset`
- [ ] unit tests
- [x] integration tests
- [x] necessary TSDoc comments

### CODE REVIEW

File: packages/core/src/routes/verification/index.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      const user = await findUserById(userId);
+      await verifyUserPassword(user, password);
+      const record = await verificationRecords.insert({
+        id: generateStandardId(),
+        type: VerificationType.Password,
+        userId,
+        expiresAt: new Date(Date.now() + expirationTime).valueOf(),
+        // Password verification is verified immediately
+        verifiedAt: Date.now(),
+      });
```
Comment (simeng-li) [inline packages/core/src/routes/verification/index.ts:47]: Can we reuse the verification classes defined in the experience routes? Extract those classes to a shared place.
`/me/verifications` endpoints should have very much the same logic as experience verification endpoints.

Comment (wangsijie) [reply packages/core/src/routes/verification/index.ts]: updated

File: packages/core/src/routes/experience/classes/verifications/index.ts [hunk 2]
```diff
@@ -138,3 +140,41 @@ export const buildVerificationRecord = (
     }
   }
 };
+
+export const buildUserVerificationRecordById = async (
```
Comment (simeng-li) [inline packages/core/src/routes/experience/classes/verifications/index.ts:144]: nit: Could move these two methods to a new verification library. Decouple the get and get logic from the verification class itself.

File: packages/core/src/routes/experience/classes/verifications/new-password-identity-verification.ts [hunk 3]
```diff
@@ -28,6 +28,7 @@ export type NewPasswordIdentityVerificationRecordData = {
    * For email and phone new identity registration, a `CodeVerification` record is required.
    */
   identifier: InteractionIdentifier;
+  userId?: string;
```
Comment (simeng-li) [inline packages/core/src/routes/experience/classes/verifications/new-password-identity-verification.ts:31]: verification record itself should not contain userId info?

File: packages/core/src/routes/experience/classes/verifications/password-verification.ts [hunk 4]
```diff
@@ -20,32 +20,43 @@ import { type IdentifierVerificationRecord } from './verification-record.js';
 export type PasswordVerificationRecordData = {
   id: string;
   type: VerificationType.Password;
-  identifier: InteractionIdentifier;
+  /** If user is not identified, identifier is required */
+  identifier?: InteractionIdentifier;
+  /** If user is already identified, userId is required */
+  userId?: string;
```
Comment (simeng-li) [inline packages/core/src/routes/experience/classes/verifications/password-verification.ts:26]: userId should be a unique type of identifier?  In this case, a user is unverified but can verify the identity using `userId` and `password`.  `userId` is much like `username` here.

File: packages/core/src/routes/profile/index.ts [hunk 5]
```diff
@@ -2,32 +2,45 @@ import { z } from 'zod';
...
     }),
     async (ctx, next) => {
-      ctx.body = {
-        sub: ctx.auth.id,
-      };
-      ctx.status = 200;
+      const { id: userId } = ctx.auth;
+      const { password, verificationRecordId } = ctx.guard.body;
+
+      // TODO(LOG-9947): apply password policy
```
Comment (simeng-li) [inline packages/core/src/routes/profile/index.ts:30]: Maybe you can reuse the PasswordValidator class under `experience/classes/libraries`

File: packages/core/src/routes/profile/index.ts [hunk 6]
```diff
@@ -2,32 +2,45 @@ import { z } from 'zod';
...
+  const {
+    users: { updateUserById },
+  } = queries;
+
   router.use(koaOidcAuth(provider));
 
-  // TODO: test route only, will implement a better one later
-  router.get(
-    '/profile',
+  router.post(
```
Comment (simeng-li) [inline packages/core/src/routes/profile/index.ts:20]: Don't forget to trigger the user.updated webhooks.

File: packages/core/src/routes/verification/index.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            value: userId,
+          },
+          payload: {
+            verificationId: passwordVerification.id,
+          },
+        },
+        passwordVerification.verify(password)
+      );
+
+      const { id } = passwordVerification.toJson();
```
Comment (simeng-li) [inline packages/core/src/routes/verification/index.ts:47]: just `passwordVerification.id` ?

File: packages/core/src/routes/profile/index.ts (hunk 6 above)
Comment (wangsijie) [reply packages/core/src/routes/profile/index.ts:20]: Added a TODO issue.

File: packages/core/src/routes/profile/index.ts (hunk 5 above)
Comment (wangsijie) [reply packages/core/src/routes/profile/index.ts:30]: Sure, will implement it later.

File: packages/core/src/queries/verification-records.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    },
+  });
+
+  public readonly update = buildUpdateWhereWithPool(this.pool)(VerificationRecords, true);
+
+  public readonly find = buildFindEntityByIdWithPool(this.pool)(VerificationRecords);
+
+  constructor(public readonly pool: CommonQueryMethods) {}
+
+  public findUserActiveVerificationRecordByUserId = async (userId: string, id: string) => {
```
Comment (simeng-li) [inline packages/core/src/queries/verification-records.ts:34]:
```suggestion
  public findUserActiveVerificationRecordByVerificationId = async (userId: string, id: string) => {
```

**Review by simeng-li — APPROVED** [review]
overall LGTM

**Pull request merged**